h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -739,10 +745,10 @@ private static void enlargeTables(float[][] data, int[][] rowIndex, int cols, in
                 currentCol = 0;
                 cols -= (data[currentRow].length - currentCol);
                 currentRow++;
-                data[currentRow] = new float[1 << 20];
-                rowIndex[currentRow] = new int[1 << 20];
+                data[currentRow] = malloc4f(ALLOCATED_ARRAY_LEN);
+                rowIndex[currentRow] = malloc4(ALLOCATED_ARRAY_LEN);
             } else {
-                int newLen = (int) Math.min((long) data[currentRow].length << 1L, (long) ARRAY_MAX);
+                int newLen = (int) Math.min(Math.min((long) data[currentRow].length << 1L, getFreeMemory() / 4), (long) ARRAY_MAX);
```

Comment:
This is not right in my opinion. You should not make memory allocation based on available memory because JVM & H2O memory model doesn't work like that. You want to try to allocate the memory even if it is not available - this should eventually trigger GC and your allocation should succeed.

If you feel like doubling the size of the array is wasteful. A better fix, in my opinion, is to do 2 passes over the (meta)data - first calculate the correct target size and then allocate precisely as much memory as you need. You can observe that to get the correct size of the backing arrays of DMatrix you only need the characteristics of the data itself in the dense case (eg. number of rows, number of categorical values in each column, number of columns). This is by itself enough to calculate the desired size for non-sparse data.